src/main/java/io/neonbee/entity/UnregisterEntityModelHook.java

```diff
@@ -0,0 +1,88 @@
+
+import io.neonbee.NeonBee;
+import io.neonbee.hook.Hook;
+import io.neonbee.hook.HookContext;
+import io.neonbee.hook.HookType;
+import io.neonbee.internal.cluster.ClusterHelper;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonArray;
+
+/**
+ * Hooks for unregistering verticle models.
+ */
+public class UnregisterEntityModelHook {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UnregisterEntityModelHook.class);
+
+    /**
+     * This method is called when a NeonBee instance shutdown gracefully.
+     *
+     * @param neonBee     the {@link NeonBee} instance
+     * @param hookContext the {@link HookContext}
+     * @param promise     {@link Promise} to completed the function.
+     */
+    @Hook(HookType.BEFORE_SHUTDOWN)
+    public void unregisterOnShutdown(NeonBee neonBee, HookContext hookContext, Promise<Void> promise) {
+        LOGGER.info("Unregistering models on shutdown");
+        Vertx vertx = neonBee.getVertx();
+        String nodeId = ClusterHelper.getNodeId(neonBee.getVertx());
+        unregister(vertx, nodeId).onComplete(promise)
+                .onSuccess(unused -> LOGGER.info("models unregistered successfully"))
+                .onFailure(ignoredCause -> LOGGER.error("failed to unregister models on shutdown"));
+    }
+
+    /**
+     * Unregister the entity models for the node by ID.
+     *
+     * @param vertx  The Vert.x instance
+     * @param nodeId hazelcast node ID
+     * @return Future
+     */
+    public static Future<Void> unregister(Vertx vertx, String nodeId) {
+        if (!vertx.isClustered()) {
+            return succeededFuture();
+        }
+
+        LOGGER.info("unregistering entity verticle models for node ID {} ...", nodeId);
+        Future<Void> unregisterFuture = EntityVerticle.getNodeEntitiesMap(vertx).compose(map -> map.get(nodeId))
+                .map(jsonArray -> jsonArray == null ? new JsonArray() : (JsonArray) jsonArray)
+                .compose(entries -> allComposite(EntityVerticle.removeSharedMapEntries(vertx, entries)).mapEmpty());
+
+        unregisterFuture
+                .onSuccess(unused -> LOGGER.info("unregistered entity verticle models for node ID {} ...", nodeId))
+                .onFailure(cause -> LOGGER.error("failed to unregistered entity verticle models for node ID {} ...",
+                        nodeId, cause));
+
+        return unregisterFuture;
+    }
+
+    /**
+     * This method is called when a NeonBee node has left the cluster.
+     *
+     * @param neonBee     the {@link NeonBee} instance
+     * @param hookContext the {@link HookContext}
+     * @param promise     {@link Promise} to completed the function.
+     */
+    @Hook(HookType.NODE_LEFT)
+    public void cleanup(NeonBee neonBee, HookContext hookContext, Promise<Void> promise) {
+        String nodeId = hookContext.get(CLUSTER_NODE_ID);
+        LOGGER.info("cleanup models for node {}", nodeId);
```

Comment:
You don't clean up models, you clean up QualifiedNames for Entities.

Reply:
good point!